type_lens/callable_view.py

```diff
@@ -0,0 +1,80 @@
+from __future__ import annotations
+
+import inspect
+import sys
+import types
+from typing import TYPE_CHECKING, Any, Callable
+
+from typing_extensions import get_type_hints
+
+from type_lens.parameter_view import ParameterView
+from type_lens.type_view import TypeView
+
+__all__ = ("CallableView",)
+
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+
+
+class CallableView:
+    def __init__(self, fn: Callable, type_hints: dict[str, type]):
+        self.callable = fn
+        self.signature = getattr(fn, "__signature__", None) or inspect.signature(fn)
+
+        return_annotation = type_hints.pop("return", None)
+        self.return_type = TypeView(return_annotation)
+
+        self.parameters = tuple(
+            ParameterView.from_parameter(param, type_hints) for param in self.signature.parameters.values()
+        )
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, CallableView):
+            return False
+
+        return bool(
+            self.callable == other.callable
+            and self.parameters == other.parameters
+            and self.return_type == other.return_type
+        )
+
+    def __repr__(self) -> str:
+        cls_name = self.__class__.__name__
+
+        return f"{cls_name}({self.callable.__name__})"
+
+    @classmethod
+    def from_callable(
+        cls: type[Self],
+        fn: Callable,
+        *,
+        get_type_hints: Callable = get_type_hints,
```

Comment:
Also this, i included this because of this: https://github.com/DanCardin/cappa/blob/main/src/cappa/typing.py#L197

But I could also leave this out, if you were willing to inline it. It provides:

* Use `_forward_ref` instead of `typing.ForwardRef` to handle the `is_class` argument (i forget what exactly this affects, but it's super annoying)
* Uses `eval_type_backport` instead of `eval_type`, to backport `a | b` to python 3.8/9

Reply:
It's somewhat moot for now because, even if for my own sanity writing tests in cappa, 3.9 doesn't support runtime evaluated pipe unions either, and 3.9/10/11+ are wildly different in inconvenient ways.

I dont find 3.8 versus 3.9 to be particularly different, they're both deficient in the same ways. It's once 3.9 becomes EOL that i'd probably drop support for both, because they both require the same hacks.

> This would be a nice to have, for sure. Perhaps we could drop the argument to the function, detect if eval-type-backport is installed and use the patched get_type_hints implementation in that case.

Yea i included it here because I will need to backport, but i can remove it (now, or when including the patch) if you're interested in pulling that compatibility layer over too.

Reply:
> I dont find 3.8 versus 3.9 to be particularly different, they're both deficient in the same ways. It's once 3.9 becomes EOL that i'd probably drop support for both, because they both require the same hacks.

The applicable difference here is whether we need to worry about that ForwardRef hack, fwict it just drops the is_class arg, which is only 3.8.

Reply:
The real impetus was swapping out `eval_type` for the packported version, which is relevant to 3.9 also. So to get that feature, the whole `get_type_hints` implementation needs to be swapped too. at that point, the ForwardRef hack isn't that big of an imposition, imo.